### ISSUE TITLE
Training stability paper vecfield

### ISSUE THREAD

**Issue by r1cc4r2o** [ccr-cheng/statistical-flow-matching#2]
I've been using your notebook "swissroll.ipynb" to reproduce your results on swissroll, but I noticed that when I am using the vecfield (as it is defined in the paper) the training is not stable and the model does not converge.

```python
    @classmethod
    def vecfield(cls, p, q, t, eps=0.) -> (Tensor, Tensor):
        """
        Vector field at timestep t.
        :param p: source data point, Tensor of shape (..., n)
        :param q: target data point, Tensor of shape (..., n)
        :param t: timestep between 0 and 1, Tensor of shape (...)
        :param eps: small float to avoid instability
        :return:
            pt: interpolant at timestep t, Tensors of shape (..., n)
            vf: vector field, Tensors of shape (..., n)
        """
        pt = cls.interpolate(p, q, t, eps)
        vf = cls.log(pt, q, eps) / (1 - t).unsqueeze(-1)
        return pt, vf
```

On the other hand, if I use the following definition for the vector field (defined in the class SphereCategoricalFlow), the training is stable and the model converges:

```python
    @classmethod
    def vecfield(cls, p, q, t, eps=1e-4):
        x, y = p, q
        dist = cls.dist(x, y, eps).unsqueeze(-1)
        xt = cls.interpolate(x, y, t, eps)
        ux = x - xt
        ux = ux - (xt * ux).sum(dim=-1, keepdim=True) * xt
        ux_norm = ux.norm(dim=-1, keepdim=True)
        uy = y - xt
        uy = uy - (xt * uy).sum(dim=-1, keepdim=True) * xt
        vf = dist * torch.where(ux_norm > eps, -ux / ux_norm, F.normalize(uy, dim=-1))
        return xt, vf
```

I could not find any reference to this definition in the paper, so I was wondering if you could clarify why this definition works well. Is there a specific reason for this difference?

Thank you in advance for your help. I appreciate your work and the insights you provide in your research.

Hoping to hear from you soon.

**Comment by ccr-cheng** [comment]
Hi Riccardo,

Thanks for your interest in our work. The instability of the default `vecfield` method came from the division by 1-t, which may be very close to 0 as t is uniformly randomly sampled. To avoid such instability, **you can modify the `max_t` argument of the `CategoricalFlow` class to a value less than 1** (a common choice would be 0.999). You can do this by simply adding the `max_t` field under the `model` field. The default value of `max_t` is 1, because for certain setups, such as the Euclidean assumption (`LinearCategoricalFlow`), dividing by 1-t effectively cancels out.

The spherical `vecfield` implementation is a numerically stable one that allows for `max_t` of 1. Such an implementation is a corollary of Eq.20 (the logarithm map on the sphere) in Appendix A.2 by noticing the following fact:
- Conditional vector fields along the geodesic have a constant norm: $\\|u_t\\|_g^2=\\|u_0\\|_g^2=d_g^2(x_0,x_1),\forall t\in[0,1]$.
- The spherical vector field is perpendicular to the position vector, i.e. $\langle x_t,u_t\rangle=0$.

Therefore, in our implementation, we first calculate the geodesic distance between the two points and then find the direction of the vector field by orthogonally projecting it to the tangent space. We then normalize the direction and scale it by the geodesic distance. Another trick here is to simultaneously compute the two projections in case that, if one projection is too small (say `ux = proj(x - xt)`), we can still use the projection of `uy = proj(y - xt)` (note that their directions only differ by a sign).

Hope the above clarifications can answer your questions!

Chaoran

**Comment by r1cc4r2o** [comment]
Hi Chaoran, 

Thank you very much for taking the time to clarify my questions.

Best

Riccardo
